registry/machine.go

```diff
@@ -104,3 +130,27 @@ func (r *EtcdRegistry) RemoveMachineState(machID string) error {
 	}
 	return err
 }
+
+// mergeMetadata merges the machine-set metadata with the dynamic metadata to better facilitate
+// machines leaving and rejoining a cluster.
+// Merging metadata uses the following rules:
+// - Any keys that are only in one collection are added as-is
+// - Any keys that exist in both, the dynamic value takes precence
+// - Any keys that have a zero-value string in the dynamic metadata are considered deleted
+//   and are not included in the final collection
+func mergeMetadata(machineMetadata, dynamicMetadata map[string]string) map[string]string {
+	if dynamicMetadata == nil {
+		return machineMetadata
+	}
+	for k, v := range machineMetadata {
+		if _, ok := dynamicMetadata[k]; !ok {
+			dynamicMetadata[k] = v
+		}
+	}
+	for k, v := range dynamicMetadata {
+		if v == "" {
+			delete(dynamicMetadata, k)
+		}
+	}
+	return dynamicMetadata
```

Comment:
I wasn't worried about caller confusion since it was internal to the package, but I agree still better to allocate a new map here.